### ISSUE TITLE
Eigen Ref

### ISSUE THREAD

**Issue by jmirabel** [stack-of-tasks/eigenpy#32]
I try to bind a function whose prototype is `f(Eigen::Ref)`. As far as I understand, this package only support `f(eigenpy::Ref)`. Am I right ?

**Comment by jcarpent** [comment]
eigenpy::ref is a simple inheritance from Eigen::Ref. You just need to add in your Python wrapping something like:
`
template<typename MatType>
void setOnes_wrap(eigenpy::Ref<MatType> mat)
{
  setOnes(mat);
}
`
where 

`template<typename MatType,int Options, typename StrideType>
void setOnes(Eigen::Ref<MatType,Options,StrideType> mat)
{
  mat.setOnes();
}`

I just updated the unit test related to ref aka unittest/ref.cpp.

**Comment by jcarpent** [comment]
Please see https://github.com/stack-of-tasks/eigenpy/blob/devel/unittest/ref.cpp

**Comment by jmirabel** [comment]
I does not solve my issue because it needs the library to be bound to be template, i.e. not
```cpp
void function (Eigen::Ref<Vector> mat)
```
but
```cpp
template<typename MatType,int Options, typename StrideType>
void function (Eigen::Ref<MatType,Options,StrideType> mat)
```

Can't `Eigen::Map` solve this issue ?

**Comment by jmirabel** [comment]
> Can't Eigen::Map solve this issue ?

I think it cannot...

**Comment by jcarpent** [comment]
No. This comes from the fact that Numpy is by default row major while Eigen is col major.
One way to solve this issue is by imposing Numpy to be column major. And then due similar wrapper, but without the stride option.

**Comment by jmirabel** [comment]
Don't you think there should be two converter in eigenpy (one for each storage order of numpy matrices) ? Or only accept col-major Numpy matrices (and maybe provide a function that creates a suitable Numpy matrix from any type of Numpy matrix) ?

**Comment by jcarpent** [comment]
This is the idea of eigenpy::ref to be compatible with both conventions.
For the second choice, I would prefer to avoid as it will create additional memory that are hard to track for end users.
If you don't want to share memory between numpy and Eigen, just create a wrapper of your function having Eigen::MatrixXd as inputs. It will be simpler.

The main issue comes seems to come from numpy which does not allow to select the order convention when importing numpy.

**Comment by jmirabel** [comment]
But then I do not understand the purpose of `eigenpy::Ref`. If one has to write:
```cpp
template<typename MatType,int Options, typename StrideType>
void function (Eigen::Ref<MatType,Options,StrideType> mat)
```
instead of:
```cpp
void function (Eigen::Ref<Vector> mat)
```
why would one use `Eigen::Ref` rather than `Eigen::MatrixBase` ? I mean that the purpose of `Eigen::Ref` is to avoid the burden of having to make all function taking Eigen object template. I think `eigenpy::Ref` is only useful if it works with the second prototype.

**Comment by jcarpent** [comment]
As I told you last time, this a preliminary work. 
It requires more stuff to do for final integration. 
The only thing I can tell you is do not use it if you don't need to share memory. 
Just make a wrapper it is simpler.
Le Lundi, Mai 07, 2018 10:12 CEST, Joseph Mirabel <notifications@github.com> a écrit:
 
But then I do not understand the purpose of eigenpy::Ref. If one has to write:template<typename MatType,int Options, typename StrideType>
void function (Eigen::Ref<MatType,Options,StrideType> mat)
instead of:void function (Eigen::Ref<Vector> mat)
why would one use Eigen::Ref rather than Eigen::MatrixBase ? I mean that the purpose of Eigen::Ref is to avoid the burden of having to make all function taking Eigen object template. I think eigenpy::Ref is only useful if it works with the second prototype.

**Comment by jmirabel** [comment]
Got it. Sorry for the noise then.
